app/src/adapters/db/clients/postgres_client.py

```diff
@@ -0,0 +1,149 @@
+import logging
+import os
+import urllib.parse
+from typing import Any
+
+import psycopg2
+import sqlalchemy
+import sqlalchemy.pool as pool
+
+from src.adapters.db.client import DBClient
+from src.adapters.db.config import DBConfig, get_db_config
+
+logger = logging.getLogger(__name__)
+
+
+class PostgresDBClient(DBClient):
+    """
+    An implementation of a DBClient for connecting to a Postgres DB
+    as configured by parameters passed in from the db_config
+    """
+
+    def __init__(self, db_config: DBConfig | None = None) -> None:
+        if not db_config:
+            db_config = get_db_config()
+        self.db_config = db_config
+        super().__init__()
+
+        if self.db_config.check_connection_on_init:
+            self.check_db_connection()
+
+    def _configure_engine(self) -> sqlalchemy.engine.Engine:
+        # We want to be able to control the connection parameters for each
+        # connection because for IAM authentication with RDS, short-lived tokens are
+        # used as the password, and so we potentially need to generate a fresh token
+        # for each connection.
+        #
+        # For more details on building connection pools, see the docs:
+        # https://docs.sqlalchemy.org/en/13/core/pooling.html#constructing-a-pool
+        def get_conn() -> Any:
+            return psycopg2.connect(**get_connection_parameters(self.db_config))
```

Comment:
at this point i think it's a bit more opaque to the reader of this function what the superclass constructor does, and the tricky dependency between needing to set the self.db_config instance variable before calling super().__init__() which then calls _configure_engine which relies on the side effect of self.db_config already being set, and itself having a side effect of setting self._engine, feels like a bit too much reliance on side effects.

at this point i think it'd be cleaner to remove `__init__` and `_configure_engine` from the DBClient abstract base class and instead just make it clear from the typing system that self._engine needs to be set in the constructor (i.e. it has no default value and is not an Optional type and therefore needs to be set) i.e.

```python
class DBClient(ABC):
  # Concrete subclasses need to set the _engine instance var in the constructor.
  # It should not be None.
  _engine: sqlalchemy.engine.Engine
```

then here we can remove all other reliance on side effects e.g.
```suggestion
    def __init__(self, db_config: DBConfig | None = None) -> None:
        if not db_config:
            db_config = get_db_config()
        
        self._engine = self._configure_engine(db_config)

        if db_config.check_connection_on_init:
            check_db_connection()

    def _configure_engine(self, db_config) -> sqlalchemy.engine.Engine:
        # We want to be able to control the connection parameters for each
        # connection because for IAM authentication with RDS, short-lived tokens are
        # used as the password, and so we potentially need to generate a fresh token
        # for each connection.
        #
        # For more details on building connection pools, see the docs:
        # https://docs.sqlalchemy.org/en/13/core/pooling.html#constructing-a-pool
        def get_conn() -> Any:
            return psycopg2.connect(**get_connection_parameters(db_config))
```

Reply:
Let me see if typing allows that, otherwise I'll just make the init function take in an engine that the subclass is expected to handle.

Reply:
Actually was perfectly fine to do that, so that does seem a bit cleaner